components/post_view/post_list.jsx

```diff
@@ -173,24 +178,61 @@ export default class PostList extends React.PureComponent {
     }
 
     static getDerivedStateFromProps(props, state) {
-        const postListIds = props.postListIds || [];
-        let newPostListIds;
-
-        if (state.atEnd) {
-            newPostListIds = [...postListIds, PostListRowListIds.CHANNEL_INTRO_MESSAGE];
-        } else if (props.postVisibility >= Constants.MAX_POST_VISIBILITY) {
-            newPostListIds = [...postListIds, PostListRowListIds.MAX_MESSAGES_LOADED];
-        } else if (state.autoRetryEnable) {
-            newPostListIds = [...postListIds, PostListRowListIds.MORE_MESSAGES_LOADER];
-        } else {
-            newPostListIds = [...postListIds, PostListRowListIds.MANUAL_TRIGGER_LOAD_MESSAGES];
+        const {prevPropsPostListIds, postListIds: statePostListIds} = state;
+        const propsPostListIds = props.postListIds;
+        let postListIds = props.postListIds || [];
+
+        if (propsPostListIds) {
+            const hasPostListIdsChanged = prevPropsPostListIds && (prevPropsPostListIds !== propsPostListIds);
+
+            if (!prevPropsPostListIds) {
+                const sliceIndexForPosts = props.focusedPostId ? MAXIMUM_POSTS_FOR_SLICING.permalink : MAXIMUM_POSTS_FOR_SLICING.channel;
+                postListIds = postListIds.slice(0, sliceIndexForPosts);
+            } else if (hasPostListIdsChanged) {
+                if (statePostListIds[0] !== propsPostListIds[0]) {
+                    // Post added at bottom
+                    const oldestPostId = PostList.getOldestVisiblePostId(statePostListIds);
```

Comment:
This isn't going to work if the last post in the channel is a combined system message. The post will have an ID like `user-activity-post1_post2`, but `getLastPostId` will just return `post2` then, so the `findIndex` call will return -1. But even if we change it so that it would get `user-activity-post1_post2`, that might not work because the previous post might be a join/leave post, so its ID would have changed to `user-activity-post1_post2_post3`.

I'm not sure how we could work around this since the combined system messages really make this difficult. Maybe we could do this slicing before the posts get combined by `preparePostIdsForPostList`. I've also thought about how we could try using timestamps here instead of post IDs since those never change, but we'd need more information about the posts to do that.

Reply:
@hmhealey Wouldn't the postID change if combined system has a new system post?. If new post is added the oldest postId from the system message changes so this should work IMO. If that is not the case then we can fix it by making sure we get the right postID from the system messages which would change if a new post is added or removed(Don't think there is a chance of this happening).

I am not in favour of the changes here. Problem with this PR as the unknowns which can cause hard to catch bugs with vague reports and this makes it really complicated for bi-directional scrolling.

I have a better solution for this. I will submit it first before working on other PR's